### PR TITLE
job: Implement job worker package

### PR DESCRIPTION
**Pull request by juliaogris**
Implement the `job` package to manage jobs, defined as processes in any
execution state with optional resource limits.

The `job.Controller` provides methods to:

- `Start`: Create and start a new job.
- `Stop`:  Stop a running job.
- `Status`: Get the status of a job.

Additionally there is a `StopAll` method to stop all running jobs and 
clean up the controller's resources.

Access control is implemented to ensure that started jobs are only
accessible by their owner. The `Status` method returns
a copy of the `job.Status` to prevent accidental modification of
job state.

Log streaming is intentionally omitted in this pull request and will be
implemented after integrating the `job` package with the API server.

### CODE REVIEW

File: pkg/job/controller.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+			return err
+		}
+	}
+	for _, ioLimit := range limits.IO {
+		if err := writeCgroupFile(cgroup, "io.max", ioLimit); err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
Comment (smallinsky) [inline pkg/job/controller.go]: Can we clean up the groups if any of the write subcalls fail?

File: pkg/job/controller.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+// get retrieves a job from the controller by ID. It is synchronized to ensure
+// safe concurrent access to the job map. It also verifies that the given owner
+// has access to the requested job.
+func (c *Controller) get(owner, id string) (*job, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	job, ok := c.jobs[id]
+	if !ok {
+		return nil, fmt.Errorf("%w: %q", ErrJobNotFound, id)
```
Comment (smallinsky) [inline pkg/job/controller.go:190]: > The returned error includes the file and line number of the caller when formatted with additional detail enabled. If the last argument is an error the returned error's Format method will return it if the format string ends with ": %s", ": %v", or ": %w". If the last argument is an error and the format string ends with ": %w", the returned error implements an Unwrap method returning it.

From: https://pkg.go.dev/fmt#Errorf

The ": %w" format is required to leverage the ErrJobNotFound type error, allowing the error to be checked using errors.Is. Otherwise, the custom error type would not be preserved.

File: pkg/job/controller.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+// Start starts a new job with the given command and arguments for the given
+// owner. It returns the ID of the newly started job, or an error if the job
+// could not be started.
+//
+// The job is executed within its own cgroup, with resource limits applied as
+// configured on the controller.
+func (c *Controller) Start(owner string, command string, args ...string) (string, error) {
+	if len(command) == 0 {
+		return "", fmt.Errorf("%w: empty command", ErrCommand)
```
Comment (smallinsky) [inline pkg/job/controller.go:90]: > The returned error includes the file and line number of the caller when formatted with additional detail enabled. If the last argument is an error the returned error's Format method will return it if the format string ends with ": %s", ": %v", or ": %w". If the last argument is an error and the format string ends with ": %w", the returned error implements an Unwrap method returning it.

From: https://pkg.go.dev/fmt#Errorf

The ": %w" format is required to leverage the ErrJobNotFound type error, allowing the error to be checked using errors.Is. Otherwise, the custom error type would not be preserved.

File: pkg/job/controller.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+	}
+
+	if c.isShutDown() {
+		return "", fmt.Errorf("cannot start command: %w", ErrShutdown)
+	}
+	id := strconv.FormatUint(c.maxID.Add(1), 10)
```
Comment (smallinsky) [inline pkg/job/controller.go:96]: Is it safe to use IDs as a sequence of numbers?

File: pkg/job/controller.go (hunk 3 above)
Comment (nklaassen) [reply pkg/job/controller.go:90]: @smallinsky your link doesn't include that text and I'm pretty sure it's not right? I just tried it out, unwrap seems to work no matter where the `%w` is.

I only found the text you quoted on some help for golang.org/x/exp/errors but we're talking about `fmt.Errorf` here

File: pkg/job/job.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+	if !j.status.Running {
+		slog.Info("job already stopped", "id", j.status.ID)
+		return nil
+	}
+	// Write "1" to <job-cgroup>/cgroup.kill to kill job process and all its children
+	if err := writeCgroupFile(j.cgroup, "cgroup.kill", "1"); err != nil {
+		return fmt.Errorf("%w: cannot write to cgroup.kill %q: %q, %w", ErrJobStop, j.status.ID, j.cgroup, err)
+	}
```
Comment (nklaassen) [inline pkg/job/job.go]: would it be possible for the job to fork some child processes, then exit itself so status is not running, but then you never write to the cgroup.kill file and the children keep running?

File: pkg/job/job.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+// wait waits for the job to finish, updates the job status and deletes its
+// cgroups. It must only be called once per job.
+func (j *job) wait() {
+	waitErr := j.cmd.Wait()
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+	j.status.Running = false
+	j.status.Stopped = time.Now()
+	exitErr := &exec.ExitError{}
```
Comment (nklaassen) [inline pkg/job/job.go]: nit: there's no need to actually allocate this error
```suggestion
	var exitErr *exec.ExitError
```

File: pkg/job/controller_test.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+	require.NoError(t, err)
+
+	status, err := controller.Status("owner", id)
+	require.NoError(t, err)
+	require.Equal(t, id, status.ID)
+	require.Equal(t, "sleep", status.Command)
+	require.Equal(t, []string{"10"}, status.Args)
+	require.Equal(t, time.Time{}, status.Stopped)
+	require.True(t, status.Running)
+	require.Equal(t, job.NotTerminated, status.ExitCode)
```
Comment (nklaassen) [inline pkg/job/controller_test.go]: nit: rather than individually comparing each field like this, I usually prefer [full structure comparisons](https://google.github.io/styleguide/go/decisions#full-structure-comparisons), I find they are easier to write and it's more likely for the test to fail if you add a new field with a non-zero value

File: pkg/job/controller.go (hunk 3 above)
Comment (smallinsky) [reply pkg/job/controller.go:90]: My fault I was mistakenly convinced that this functioned similarly. Marking this comment as resolved.

File: pkg/job/controller.go (hunk 1 above)
Comment (juliaogris) [reply pkg/job/controller.go]: Done.

File: pkg/job/controller.go (hunk 4 above)
Comment (juliaogris) [reply pkg/job/controller.go:96]: I agree to be cautious with sequential IDs. Potential concerns:

- **Enumeration attacks:** It becomes easy to guess valid job IDs (e.g., anything below the current ID).
- **Information leakage:** Sequential IDs can reveal the number of jobs started, potentially exposing usage patterns.
- **Scalability limitations:** If we scale to multiple `telejob` instances behind a load balancer, sequential IDs will no longer be unique.

In our current context, with strong authentication, a high-trust model, and no immediate scalability concerns, these risks are mitigated. However, if any of these factors change in the future, the sequential ID approach could become problematic.

To address this proactively, we could use a UUID generator like
`github.com/google/uuid.New()`. I initially opted to avoid third-party
libraries, but I'm happy to switch to UUIDs if that's preferred.

File: pkg/job/job.go (hunk 5 above)
Comment (juliaogris) [reply pkg/job/job.go]: Yes, you are right.

I have rearranged the code so that `cgroup.kill` is written inside the `job.wait()` 
method which is called independently of a job terminating naturally or by signal.
I've replaced the `cgroup.kill` write in the `job.stop()` method with 
`job.cmd.Process.Kill()`, which could be replaced by a more graceful termination
 sequence.

I could not find any definitive documentation on whether it is safe to delete a cgroup
immediately after writing to `cgroup.kill`. Some docs mention that the KILL signal to all
child processes is sent atomically, and none of my stress tests failed on an immediate
delete. I have therefore opted to just log an error if the delete fails, rather than
implement a retry after waiting a short period.

File: pkg/job/job.go (hunk 6 above)
Comment (juliaogris) [reply pkg/job/job.go]: Done.

File: pkg/job/controller_test.go (hunk 7 above)
Comment (juliaogris) [reply pkg/job/controller_test.go]: Done. Thank you!

File: pkg/job/controller.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+// newJobCgroup creates a new cgroup for a job with the specified resource
+// limits. The new cgroup is created as a subcgroup under the given parent
+// cgroup. It configures CPU, memory, and I/O limits based on the provided
+// Limits.
+func newJobCgroup(cgroup string, limits Limits) error {
+	if err := os.Mkdir(cgroup, 0o750); err != nil {
+		return fmt.Errorf("cannot create new job cgroup %q: %w", cgroup, err)
+	}
+	var err error
+	defer deleteCgroupOnErr(cgroup, err)
```
Comment (nklaassen) [inline pkg/job/controller.go]: Unfortunately this doesn't quite work, try out this [example](https://go.dev/play/p/mfw1gbzrnfu) in the go playground
```
func main() {
	var err error
	defer fmt.Println(err)
	defer func() { fmt.Println(err) }()
	err = fmt.Errorf("hello")
}
```

File: pkg/job/controller.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+		defer c.wg.Done()
+		job.wait()
+	}()
+	return id, nil
+}
+
+// Stop stops the job with the given id.
+//
+// It terminates the job's process and all its child processes by sending a
+// SIGKILL signal via the job's cgroup.kill file.
```
Comment (nklaassen) [inline pkg/job/controller.go]: This comment is now outdated

File: pkg/job/job.go (hunk 5 above)
Comment (nklaassen) [reply pkg/job/job.go]: good enough for me!

File: pkg/job/controller.go (hunk 8 above)
Comment (juliaogris) [reply pkg/job/controller.go]: It is now fixed in a second fixup commit.
Thank you!

File: pkg/job/controller.go (hunk 9 above)
Comment (juliaogris) [reply pkg/job/controller.go]: Thanks! Fixed in 1490cd3.

File: pkg/job/controller.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+// newJobCgroup creates a new cgroup for a job with the specified resource
+// limits. The new cgroup is created as a subcgroup under the given parent
+// cgroup. It configures CPU, memory, and I/O limits based on the provided
+// Limits.
+func newJobCgroup(cgroup string, limits Limits) error {
+	if err := os.Mkdir(cgroup, 0o750); err != nil {
+		return fmt.Errorf("cannot create new job cgroup %q: %w", cgroup, err)
+	}
+	var err error
+	defer func() { deleteCgroupOnErr(cgroup, err) }()
```
Comment (nklaassen) [inline pkg/job/controller.go]: This is one of the few cases where I do suggest a named return param, to be defensive. You got everything right here but it would be very easy for someone to add an `if err := writeCgroupFile(...); err != nil { return err }` later on without noticing that the cleanup wouldn't run
```suggestion
func newJobCgroup(cgroup string, limits Limits) (err error) {
	if err := os.Mkdir(cgroup, 0o750); err != nil {
		return fmt.Errorf("cannot create new job cgroup %q: %w", cgroup, err)
	}
	defer func() { deleteCgroupOnErr(cgroup, err) }()
```

**Review by nklaassen — APPROVED** [review]

File: pkg/job/controller.go (hunk 10 above)
Comment (juliaogris) [reply pkg/job/controller.go]: Right, I see, this is subtle. Fixed in 0924a3f.

**Review by smallinsky — APPROVED** [review]
